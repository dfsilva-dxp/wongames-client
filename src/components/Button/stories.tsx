import { Story, Meta } from "@storybook/react/types-6-0";
import { AddShoppingCart } from "@styled-icons/material-outlined/AddShoppingCart";
import { ComponentProps } from "react";

import Button from ".";

export default {
  title: "Form/Button",
  component: Button,
  argTypes: {
    children: {
      type: "string"
    },
    size: {
      options: ["small", "medium", "large"],
      control: { type: "select" }
    },
    icon: { type: "" }
  }
} as Meta;

export const Default: Story<ComponentProps<typeof Button>> = ({ ...props }) => (
  <Button {...props} />
);

Default.args = {
  children: "Buy now",
  size: "medium",
  fullWidth: false,
  icon: "",
  minimal: false
};

export const WithIcon: Story<ComponentProps<typeof Button>> = ({
  ...props
}) => <Button {...props} />;

WithIcon.args = {
  size: "medium",
  children: "Buy now",
  icon: <AddShoppingCart />,
  fullWidth: false,
  minimal: false
};

export const Minimal: Story<ComponentProps<typeof Button>> = ({ ...props }) => (
  <Button {...props} />
);

Minimal.args = {
  size: "medium",
  children: "Buy now",
  icon: <AddShoppingCart />,
  fullWidth: false,
  minimal: true
};

export const AsLink: Story<ComponentProps<typeof Button>> = ({ ...props }) => (
  <Button {...props} />
);

AsLink.args = {
  size: "medium",
  children: "Buy now",
  icon: "",
  fullWidth: false,
  minimal: false,
  as: "a",
  href: "/link"
};
