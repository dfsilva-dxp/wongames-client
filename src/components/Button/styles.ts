import styled, { css, DefaultTheme } from "styled-components";
import { darken } from "polished";

import { ButtonProps } from "./button";

export type WrapperProps = {
  hasIcon: boolean;
} & Pick<ButtonProps, "size" | "fullWidth" | "minimal">;

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
    font-size: ${theme.font.sizes.small};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
    font-size: ${theme.font.sizes.medium};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  minimal: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};

    &:hover {
      color: ${darken(0.1, theme.colors.primary)};
    }
  `,
  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      filter: saturate(30%);
    }
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth, minimal, hasIcon, disabled }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
    color: ${theme.colors.white};
    border: 0;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    cursor: pointer;
    text-decoration: none;

    ${!!size && wrapperModifiers[size](theme)}
    ${!!fullWidth && wrapperModifiers.fullWidth()}
    ${!!minimal && wrapperModifiers.minimal(theme)}
    ${!!hasIcon && wrapperModifiers.withIcon(theme)}
    ${!!disabled && wrapperModifiers.disabled}

    &:hover {
      background: ${minimal
        ? "none"
        : `linear-gradient(180deg, #e35565 0%, #d958a6 50%)`};
    }
  `}
`;
