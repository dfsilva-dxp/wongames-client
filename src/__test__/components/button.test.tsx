import { screen, waitFor } from "@testing-library/react";
import { AddShoppingCart } from "@styled-icons/material-outlined/AddShoppingCart";

import { renderWithTheme } from "utils/tests/helpers";

import Button from "components/Button";

describe("<Button />", () => {
  it("should render the medium button by default", () => {
    const { container } = renderWithTheme(<Button>Buy now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      height: "4rem",
      padding: "0.8rem 3.2rem",
      "font-size": "1.4rem"
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the small button when size small is passed", () => {
    renderWithTheme(<Button size="small">Buy now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      height: "3rem",
      "font-size": "1.2rem"
    });
  });

  it("should render the large button when size large is passed", () => {
    renderWithTheme(<Button size="large">Buy now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      height: "5rem",
      padding: "0.8rem 4.8rem",
      "font-size": "1.6rem"
    });
  });

  it("should render the full width button when fullWidth option is passed", () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      width: "100%"
    });
  });

  it("should render an icon", () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>
    );
    expect(screen.getByText(/Buy now/i)).toBeInTheDocument();
    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should render a minimal", () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />} minimal>
        Buy now
      </Button>
    );

    expect(screen.getByRole("button", { name: /buy now/i })).toHaveStyle({
      background: "none",
      color: "#F231A5"
    });

    waitFor(() => {
      expect(screen.getByRole("button", { name: /buy now/i })).toHaveStyleRule(
        "background",
        "none",
        {
          modifier: ":hover"
        }
      );
    });
  });

  it("should render a disabled Button", () => {
    renderWithTheme(<Button disabled>Buy now</Button>);

    waitFor(() => {
      expect(screen.getByRole("button", { name: /buy now/i })).toHaveStyleRule(
        "cursor",
        "not-allowed",
        {
          modifier: ":disabled"
        }
      );
    });
  });

  it("should render button as a link", () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>
    );

    expect(screen.getByRole("link", { name: /buy now/i })).toHaveAttribute(
      "href",
      "/link"
    );
  });
});
