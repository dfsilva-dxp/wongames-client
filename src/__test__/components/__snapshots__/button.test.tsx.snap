// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium button by default 1`] = `
<button
  class="sc-beqWaB dyXRan"
>
  <span>
    Buy now
  </span>
</button>
`;
